diff --git a/include/ann_exception.h b/include/ann_exception.h
index 6b81373c..4d9d5887 100644
--- a/include/ann_exception.h
+++ b/include/ann_exception.h
@@ -2,9 +2,11 @@
 // Licensed under the MIT license.
 
 #pragma once
+#include <cstdint>
 #include <string>
 #include <stdexcept>
 #include <system_error>
+#include <cstdint>
 #include "windows_customizations.h"
 
 #ifndef _WINDOWS
diff --git a/include/utils.h b/include/utils.h
index d3af5c3a..a9ebb89e 100644
--- a/include/utils.h
+++ b/include/utils.h
@@ -2,6 +2,7 @@
 // Licensed under the MIT license.
 
 #pragma once
+#include <xmmintrin.h>
 
 #include <errno.h>
 
diff --git a/src/index.cpp b/src/index.cpp
index 486d41e7..4d31a823 100644
--- a/src/index.cpp
+++ b/src/index.cpp
@@ -4,6 +4,7 @@
 #include <omp.h>
 
 #include <type_traits>
+#include <vector>
 
 #include "boost/dynamic_bitset.hpp"
 #include "index_factory.h"
@@ -25,6 +26,13 @@
 
 #define MAX_POINTS_FOR_USING_BITSET 10000000
 
+#include "vamana_instr.h"
+
+std::vector<VamanaHopEntry> vamana_hop_log;
+bool vamana_instr_enabled = false;
+thread_local int vamana_current_qid = -1;
+
+
 namespace diskann
 {
 // Initialize an index with metric m, load the data of type T with filename
@@ -961,7 +969,15 @@ std::pair<uint32_t, uint32_t> Index<T, TagT, LabelT>::iterate_to_fixed_point(
         }
 
         assert(dist_scratch.capacity() >= id_scratch.size());
-        compute_dists(id_scratch, dist_scratch);
+        
+	// Batching opportunity decay measurement
+        if (vamana_instr_enabled) {
+            for (auto id : id_scratch) {
+                vamana_hop_log.push_back({vamana_current_qid, (int)hops, (int)id});
+            }
+        }
+	hops++;
+	compute_dists(id_scratch, dist_scratch);
         cmps += (uint32_t)id_scratch.size();
 
         // Insert <id, dist> pairs into the pool of candidates
diff --git a/include/vamana_instr.h b/include/vamana_instr.h
new file mode 100644
index 00000000..01f71b1c
--- /dev/null
+++ b/include/vamana_instr.h
@@ -0,0 +1,12 @@
+#pragma once
+#include <vector>
+
+struct VamanaHopEntry {
+    int query_id;
+    int hop;
+    int node_id;
+};
+
+extern std::vector<VamanaHopEntry> vamana_hop_log;
+extern bool vamana_instr_enabled;
+extern thread_local int vamana_current_qid;
